File: AllSpice/Services/RecipesService.cs
namespace AllSpice.Services;
public class RecipesService
{
  private readonly RecipesRepository _recipesRepo;
  private readonly FavoritesRepository _favoritesRepo;

  public RecipesService(RecipesRepository recipesRepo, FavoritesRepository favoritesRepo)
  {
    _recipesRepo = recipesRepo;
    _favoritesRepo = favoritesRepo;
  }

  internal Recipe CreateRecipe(Recipe newRecipe)
  {
    return _recipesRepo.CreateRecipe(newRecipe);
  }

  internal List<Recipe> GetAllRecipes()
  {
    return _recipesRepo.GetAllRecipes();
  }

  internal Recipe GetById(int recipeId)
  {
    Recipe foundRecipe = _recipesRepo.GetById(recipeId);

    if (foundRecipe == null)
    {
      throw new Exception("Recipe does not exist");
    }
    return foundRecipe;
  }

  internal void RemoveRecipe(int recipeId, string accountId)
  {
    Recipe foundRecipe = GetById(recipeId);
    if (foundRecipe == null)
    {
      throw new Exception("Recipe not found");
    }
    if (recipeId == 0)
    {
      throw new Exception("This Recipe doesn't exist");
    }
    if (foundRecipe.CreatorId != accountId)
    {
      throw new Exception("Unauthorized to delete recipe");
    }

    foundRecipe.Archived = true;
    _recipesRepo.RemoveRecipe(foundRecipe);
  }

  internal Recipe EditRecipe(Recipe recipeData, string accountId)
  {

    if (recipeData.CreatorId != accountId)
    {
      throw new Exception("Unauthorized to edit this recipe");
    }

    Recipe original = GetById(recipeData.Id);
    // original.Category = recipeData.Category ?? original.Category;
    // original.Img = recipeData.Img ?? original.Img;
    original.Instructions = recipeData.Instructions ?? original.Instructions;
    // original.Title = recipeData.Title ?? original.Title;
    // original.Creator=original.Creator;
    // original.CreatorId=original.CreatorId;
    // original.FavoriteCount=original.FavoriteCount;
    // original.Archived=original.Archived;

    Recipe recipe = _recipesRepo.EditRecipe(original);
    return recipe;
  }
}

File: AllSpice.client/src/services/RecipesService.js
import { AppState } from "../AppState.js";
import { Recipe } from "../models/Recipe.js";
import { api } from "./AxiosService.js"


class RecipesService {
  async getAllRecipes() {
    const res = await api.get('api/recipes')
    // console.log(res.data);
    AppState.recipes = res.data.map(r => new Recipe(r), ...AppState.recipes)
    // console.log(AppState.recipes);
  }

  async getRecipeDetails(recipeId) {
    const res = await api.get(`api/recipes/${recipeId}`)
    // console.log(res.data);
    AppState.activeRecipe = new Recipe(res.data)
    console.log(AppState.activeRecipe);
  }

  async createRecipe(recipeData) {
    const res = await api.post("api/recipes", recipeData);
    console.log("[createRecipe]", res.data);
    const newRecipe = new Recipe(res.data);
    AppState.recipes = [newRecipe, ...AppState.recipes];
  }

  async removeRecipe(recipeId) {
     await api.delete(`api/recipes/${recipeId}`)
    AppState.recipes = AppState.recipes.filter(r => r.id != recipeId)
  }

  async getAccountRecipes() {
    const res = await api.get("api/recipes");
    console.log(res.data);
    AppState.recipes = res.data.map((r) => new Recipe(r));
   AppState.recipes = AppState.recipes.filter((r) => r.creator.id == AppState.account.id);
  }

  async getRecipesBySearchTerm(query) {
    const res = await api.get("api/recipes")
    AppState.recipes = res.data.map(r => new Recipe(r))
    // AppState.recipes = AppState.recipes.filter(f => f.category.includes(query))
    
    AppState.recipes = AppState.recipes.filter((f) =>
      f.title.toUpperCase().includes(query.toUpperCase())
    );
  }
}

export const recipesService = new RecipesService()